src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenCallTests.cs

```diff
@@ -10384,38 +10383,201 @@ .maxstack  4
                 T V_4)
   IL_0000:  ldarg.0
   IL_0001:  call       ""int Program.GetArrayIndex()""
-  IL_0006:  ldelema    ""T""
-  IL_000b:  stloc.1
-  IL_000c:  ldloca.s   V_4
-  IL_000e:  initobj    ""T""
-  IL_0014:  ldloc.s    V_4
-  IL_0016:  box        ""T""
-  IL_001b:  brtrue.s   IL_0028
-  IL_001d:  ldloc.1
-  IL_001e:  ldobj      ""T""
-  IL_0023:  stloc.0
-  IL_0024:  ldloca.s   V_0
-  IL_0026:  br.s       IL_0029
-  IL_0028:  ldloc.1
-  IL_0029:  ldarga.s   V_0
-  IL_002b:  call       ""int Program.GetOffset<T>(ref T[])""
-  IL_0030:  stloc.2
-  IL_0031:  dup
-  IL_0032:  ldloc.2
-  IL_0033:  constrained. ""T""
-  IL_0039:  callvirt   ""int IMoveable.this[int].get""
-  IL_003e:  stloc.3
-  IL_003f:  ldloc.2
-  IL_0040:  ldloc.3
-  IL_0041:  ldc.i4.1
-  IL_0042:  add
-  IL_0043:  constrained. ""T""
-  IL_0049:  callvirt   ""void IMoveable.this[int].set""
-  IL_004e:  ret
+  IL_0006:  readonly.
```

Comment:
I'm a bit warry of generating unverifiable IL. It's necessary sometimes, particularly with newer constructs that were created after verification was frozen. That doesn't seem to be the case for this. 

I'm not as familiar with the `.readonly` prefix and had to look it up for this review. The set of [allowed instructions][1] does not include `stloc / ldloc`. Wondering if that is an ommision or not. The subsequent operations we perform after `ldloc` seem to fall into the allowed set.

I'd prefer to avoid this if possible. This is the opcodes that we generate today for the method: 

```il
        .maxstack 8

        IL_0000: nop
        IL_0001: ldarg.0
        IL_0002: ldc.i4.0
        IL_0003: ldelema !!T
        IL_0008: dup
        IL_0009: constrained. !!T
        IL_000f: callvirt instance int32 I1::get_P()
        IL_0014: ldc.i4.2
        IL_0015: add
        IL_0016: constrained. !!T
        IL_001c: callvirt instance void I1::set_P(int32)
        IL_0021: nop
        IL_0022: ret
```

Where is the `ArrayTypeMismatchException` being generated today? Is it on the `ldelema` instruction because the the type of the array is `Base` but the elements are `Derived`?

@davidwrighton in case he has any thoughts on how to best approach this.

[1]: https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.readonly?view=net-7.0